src/main/cljc/jarl/builtins/http.cljc

```diff
@@ -0,0 +1,199 @@
+    308 "308 Permanent Redirect"
+    400 "400 Bad Request"
+    402 "402 Payment Required"
+    404 "404 Not Found"
+    405 "405 Method Not Allowed"
+    406 "406 Not Acceptable"
+    409 "409 Conflict"
+    410 "410 Gone"
+    411 "411 Length Required"
+    412 "412 Precondition Failed"
+    413 "413 Payload Too Large"
+    414 "414 URI Too Long"
+    416 "416 Range Not Satisfiable"
+    417 "417 Expectation Failed"
+    425 "425 Too Early"
+    426 "426 Upgrade Required"
+    428 "428 Precondition Required"
+    429 "429 Too Many Requests"
+    431 "431 Request Header Fields Too Large"
+    451 "451 Unavailable For Legal Reasons"
+    500 "500 Internal Server Error"
+    501 "501 Not Implemented"
+    502 "502 Bad Gateway"
+    503 "503 Service Unavailable"
+    504 "504 Gateway Timeout"
+    505 "505 HTTP Version Not Supported"
+    506 "506 Variant Also Negotiates"
+    510 "510 Not Extended"
+    511 "511 Network Authentication Required"
+    (throw (ex-info "unknown http status code" {:code code}))))
+
+(def allowed-keys
+  #{"method"
+    "url"
+    "body"
+    "enable_redirect"
+    "force_json_decode"
+    "force_yaml_decode"
+    "headers"
+    "raw_body"
+    "tls_use_system_certs"
+    "tls_ca_cert"
+    "tls_ca_cert_file"
+    "tls_ca_cert_env_variable"
+    "tls_client_cert"
+    "tls_client_cert_file"
+    "tls_client_cert_env_variable"
+    "tls_client_key"
+    "tls_client_key_file"
+    "tls_client_key_env_variable"
+    "tls_insecure_skip_verify"
+    "tls_server_name"
+    "timeout"
+    "cache"
+    "force_cache"
+    "force_cache_duration_seconds"
+    "raise_error"
+    "caching_mode"})
+
+(def required-keys #{"method" "url"})
+
+(defn- format-set [s]
+  (str (str/join ", " (map #(str "\"" % "\"") s))))
+
+; TODO: verify correct type of each provided value
+(defn validate-request [request]
+  (let [keys (set (keys request))
+        invalid-keys (set/difference keys allowed-keys)
+        missing-keys (set/difference required-keys keys)
+        raise-error (get request "raise_error")]
+    (when (seq invalid-keys)
+      (throw (errors/type-ex "http.send: operand 1 invalid request parameters(s): {%s}", (format-set invalid-keys))))
+    (when (seq missing-keys)
+      (throw (errors/type-ex "http.send: operand 1 missing required request parameters(s): {%s}", (format-set missing-keys))))
+    (when (and (some? raise-error) (not (boolean? raise-error)))
+      (throw (errors/type-ex "http.send: invalid value for raise_error field")))))
+
+(defn verify-url [s]
+  (if (nil? (re-find #"https?://(\S+)" s))
+    ; note: does not mimic OPA error messages yet
+    (throw (errors/builtin-ex "http.send: invalid URL: %s" s))
+    s))
+
+(defn- to-json [x]
+  (if (string? x)
+    x
+    (json/write-str x)))
+
+(def ^:private unit->nano-map
+  {"h"  3600000000000
+   "m"  60000000000
+   "s"  1000000000
+   "ms" 1000000
+   "us" 1000
+   "µs" 1000
+   "ns" 1})
+
+(defn time-unit->nanos [[_ time unit :as time-unit]]
```

Comment:
We're dropping the first list element because that is the complete match in the regex result at the call-site? Can we instead drop it there, and keep this function agnostic to how these args were created?

Reply:
That's a good point. I don't think this function was meant to be public though :) I'll take a look an will change one way or other.

Reply:
Fixed as you suggested.